### PR TITLE
fix(cli/import_map): Don't statically error on dynamic unmapped bare specifiers

### LINKED ISSUE #10168: v1.8.3 - Dynamic import that can't execute causes error when using import maps.
```js
if (never) {
  await import('whatever')
}
```
I get the `unmapped bare specifier` error. But this should not cause any error or else it will be impossible to write code which supports several runtimes.

Basically I want to check if it's running in Node.js and then to load a Node.js specific module ONLY in that case. When running in Deno or the browser then I don't need to do this, hence Deno should not complain about the code in the block which is never executed.

### LINKED ISSUE #10615: Improve the error message about the unmapped bare specifier in import maps
Currently the error message is just:
```
nickolay@frontier:~/workspace/Bryntum/siesta-monorepo/siesta$ deno run -A --unstable --quiet --import-map import_map.json tests/index.js
error: Unmapped bare specifier "path"
```
Probably should include the source position, which is the cause of the error.

### LINKED ISSUE #10616: Adding `--import-map` breaks the code
I have code that works fine without import maps:
```
nickolay@frontier:~/workspace/Bryntum/siesta-monorepo/siesta$ deno run -A --unstable --quiet tests/hook/hook.t.js
Launching test file: tests/hook/hook.t.js in Deno 1.9.2

 PASS  tests/hook/hook.t.js

Test files : 1 passed, 0 failed, 1 total
Time       : 33ms
```
Adding a dummy import map breaks it:
```
nickolay@frontier:~/workspace/Bryntum/siesta-monorepo/siesta$ deno run -A --unstable --quiet --import-map import_map.json tests/hook/hook.t.js
error: Unmapped bare specifier "path"
```

How is this possible? It seems adding a source map changes the set of files being imported somehow. I do have the bare "path" but only in files that are used for Node. And again, the code works fine w/o import maps.

Import map content for the full picture:
```
{
  "imports":  {
    "siesta/":  "./"
  }
}
```

To reproduce:
1) clone the `git@github.com:bryntum/siesta.git`, specific revision is: 6e7e2576e0afcdd3f4f6e1966923ea46c29896eb
2) `npm i` , `npx tsc`
3) `cd examples/deno`
4) Launch w/o import maps: `deno run -A --unstable --quiet --no-check tests/index.ts`
Should see something like:
```
Launching test suite project: Awesome Deno project test suite in Deno 1.9.2

 FAIL  module/zoomer.t.js
├─✘ Using `zoomer` should work
│ └─✘ equal(received, expected)
│   
│     2 | import { zoomer } from "../../src/module.js";
│     3 | it('Using `zoomer` should work', async (t) => {
│   ➤ 4 |     t.equal(zoomer('Myers Courtney', 35), { name: 'Courtney Myers', age: 35 });
│       |       ^
│     5 | });
│   
│   Received                   │ │ Expected                  
│                              │ │                           
│   {                          │ │ {                         
│     "age": 35,               │ │   "age": 35,              
│     "name": "Myers Courtney" │ │   "name": "Courtney Myers"
│   }                          │ │ }                         
│   
```
5) Launch with import maps: `deno run -A --unstable --quiet --no-check --import-map import_map.json tests/index.ts`
The output is:
```
error: Unmapped bare specifier "path"
```

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #10168.
Fixes #10615.
Fixes #10616.

### CODE REVIEW

File: cli/import_map.rs [hunk 1]
```diff
@@ -454,10 +454,7 @@ impl ImportMap {
       return Ok(as_url);
     }
 
-    Err(ImportMapError(format!(
-      "Unmapped bare specifier {:?}",
-      specifier
-    )))
+    Ok(None)
```
Comment (nayeemrmn) [inline cli/import_map.rs:457]: Not spec compliant, may have to change the call sites instead.

**Comment by bartlomieju** [comment]
I'm entirely sure about this change - it seems to start deviating from the spec algorithm, but on the other hands none of the WPT fail so I guess it's okay.

**Comment by nayeemrmn** [comment]
Which part deviates from the spec algorithm? The two important changes are:
1. Structured import map errors (https://github.com/denoland/deno/pull/10618/files#diff-fba55433053f0ea5eef635f4d576f13beea8b8091c1d692f3513e46e4715e0d3L17-R20).
2. Solidifying `Module::parse()`, which doesn't error on `ModuleResolutionError::ImportPrefixMissing`, to also not error on its equivalent `ImportMapError::UnmappedBareSpecifier` (https://github.com/denoland/deno/pull/10618/files#diff-8439fae0dccaa5fecd30e137ceb78680f3b7884c1ae74a5f7c649f5273abf3e2L384-R391). This was the cause of the bug. I don't think it correlates with any spec.
 
All other changes are 1-to-1 simplifications e.g. the `ImportMap::resolve()` never returns `None`, so the `Option` is removed from its return type.

**Comment by bartlomieju** [comment]
> Which part deviates from the spec algorithm? The two important changes are:
> 
> 1. Structured import map errors (https://github.com/denoland/deno/pull/10618/files#diff-fba55433053f0ea5eef635f4d576f13beea8b8091c1d692f3513e46e4715e0d3L17-R20).
> 2. Solidifying `Module::parse()`, which doesn't error on `ModuleResolutionError::ImportPrefixMissing`, to also not error on its equivalent `ImportMapError::UnmappedBareSpecifier` (https://github.com/denoland/deno/pull/10618/files#diff-8439fae0dccaa5fecd30e137ceb78680f3b7884c1ae74a5f7c649f5273abf3e2L384-R391). This was the cause of the bug. I don't think it correlates with any spec.
> 
> All other changes are 1-to-1 simplifications e.g. the `ImportMap::resolve()` never returns `None`, so the `Option` is removed from its return type.

@nayeemrmn I see, somehow I thought spec required to return `null`/`None` when the resolution fails, but it seems you're just unifying the behavior. 

LGTM then, I will let @kitsonk take a look before landing

**Comment by kitsonk** [comment]
Sorry, for some reason this was missing from my todo list... I will look at first thing tomorrow my time.

**Review by kitsonk — APPROVED** [review]
LGTM

**Pull request merged**